File: unique_paths.py
#Time:O(m*n)
#space:O(m*n)
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        
        if m==0 and n == 0:
            return 1
        dp_matrix = [[1 for _ in range(n)] for _ in range(m)]
        for i in range(1,m):
            for j in range(1,n):
                dp_matrix[i][j] = dp_matrix[i-1][j]+dp_matrix[i][j-1]
        return dp_matrix[m-1][n-1]

File: unique-paths.cpp
//Time - O(n^2)
//Space - O(n^2)
class Solution {
public:
    int uniquePaths(int m, int n) {
        int c = 0;
        vector<vector<int>> dp(m,vector<int>(n,0));
        dp[m-1][n-1] = 1;
        for(int i=0;i<m;i++){
            dp[i][n-1] = 1;
        }
        for(int j=0;j<n;j++){
            dp[m-1][j] = 1;
        }
        for(int i=m-2;i>=0;i--){
            for(int j=n-2;j>=0;j--){
                dp[i][j] = dp[i+1][j] + dp[i][j+1];
            }
        }
        
        
        return dp[0][0];
    }
};